File: mini-uber/mini-uber.cc
/**
 * Definition of Trip:
 * class Trip {
 * public:
 *     int id; // trip's id, primary key
 *     int driver_id, rider_id; // foreign key
 *     double lat, lng; // pick up location
 *     Trip(int rider, double lat, double lng);
 * }
 * Definition of Helper
 * class Helper {
 * public:
 *     static get_distance(double lat1, double lng1,
 *                         double lat2, double lng2) {
 *         // return distance between (lat1, lng1) and (lat2, lng2)
 *     }
 * };
 */
class Location {
public:
    double lat, lng;
    Location(double _lat, double _lng): lat(_lat), lng(_lng) {}
};

class MiniUber {
public:
    MiniUber() {
        // initialize your data structure here.
        locations.clear();
        trips.clear();
    }
    
    // @param driver_id an integer
    // @param lat, lng driver's location
    // return matched trip information if there have matched rider or NULL
    Trip* report(int driver_id, double lat, double lng) {
        // Write your code here
        if (trips.find(driver_id) != trips.end()) {
            return trips[driver_id];
        }
        if (locations.find(driver_id) != locations.end()) {
            locations[driver_id]->lat = lat;
            locations[driver_id]->lng = lng;
        } else {
            locations[driver_id] = new Location(lat, lng);
        }
        return NULL;
    }
    
    // @param rider_id an integer
    // @param lat, lng rider's location
    // return a trip
    Trip* request(int rider_id, double lat, double lng) {
        // Write your code here
        double dist = -1, t_dist;
        int driver_id = -1;
        for (map<int, Location*>::iterator it = locations.begin(); it != locations.end(); ++it) {
            t_dist = Helper::get_distance(it->second->lat, it->second->lng, lat, lng);
            if (dist < 0 || dist > t_dist) {
                dist = t_dist;
                driver_id = it->first;
            }
        }
        Trip* trip = new Trip(rider_id, lat, lng);
        trip->driver_id = driver_id;
        if (driver_id != -1) {
            locations.erase(locations.find(driver_id));
            trips[driver_id] = trip;
        }
        return trip;
    }
    
private:
    map<int, Location*> locations;
    map<int, Trip*> trips;
};

File: mini-uber/mini-uber.py
'''
    Definition of Trip:
    class Trip:
    self.id; # trip's id, primary key
    self.driver_id, self.rider_id; # foreign key
    self.lat, self.lng; # pick up location
    def __init__(self, rider_id, lat, lng):
    
    Definition of Helper
    class Helper:
    @classmethod
    def get_distance(cls, lat1, lng1, lat2, lng2):
    # return calculate the distance between (lat1, lng1) and (lat2, lng2)
    '''
from Trip import Trip, Helper


class MiniUber:
    
    def __init__(self):
        # initialize your data structure here.
        self.locs = {}
        self.trips = {}
    
    # @param {int} driver_id an integer
    # @param {double} lat, lng driver's location
    # return {trip} matched trip information if there have matched rider or null
    def report(self, driver_id, lat, lng):
        # Write your code here
        if driver_id in self.trips:
            return self.trips[driver_id]
        self.locs[driver_id] = (lat, lng)
        return None
    
    
    # @param rider_id an integer
    # @param lat, lng rider's location
    # return a trip
    def request(self, rider_id, lat, lng):
        # Write your code here
        dist = -1
        driver = -1
        for key, value in self.locs.items():
            td = Helper.get_distance(value[0], value[1], lat, lng)
            if dist == -1 or td < dist:
                dist = td
                driver = key
        
        trip = Trip(rider_id, lat, lng)
        trip.driver_id = driver
        if dist != -1:
            del self.locs[driver]
            self.trips[driver] = trip
        return trip
